File: Assignments/P02/673/main.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

int main()
{
    int count;
    string sequence;
    vector<char> stack;
    bool valid;

    cin >> count;

    for (int i = 0; i < count; i++)
    {
        cin >> sequence;
        valid = true;
        for (int j = 0; j < sequence.length(); j++)
        {
            if (sequence[j] == '(' || sequence[j] == '[')
            {
                stack.push_back(sequence[j]);
            }
            else
            {
                // if popping and stack is empty
                if (stack.size() == 0)
                {
                    valid = false;
                    break;
                }
                //check matching cases
                else if (sequence[j] == ')' && stack.back() == '(')
                    stack.pop_back();
                else if (sequence[j] == ']' && stack.back() == '[')
                    stack.pop_back();
                //if doesnt match
                else
                {
                    valid = false;
                    break;
                }
            }
        }
        //after trying all cases or break
        if (valid)
            cout << "Yes\n";
        else
            cout << "No\n";
    }
}

File: Assignments/P02/673/main.py
lines = int(input())
stack = []

for _ in range(lines):
    valid = True  # set to true before input
    sequence = [char for char in input()]
    for char in sequence:
        # append for opening
        if char == '(' or char == '[':
            stack.append(char)
        else:
            # if a popping case and empty it fails
            if len(stack) == 0:
                valid = False
                break
            # pop matching
            elif char == ')' and stack[-1] == '(':
                stack.pop()
            elif char == ']' and stack[-1] == '[':
                stack.pop()
            # fail if doesnt match case
            else:
                valid = False
                break
    # after trying all cases
    if(valid and len(stack) == 0):
        print('Yes')
    else:
        print('No')
